File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
# 코드를 작성해주세요
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
parent = [0]*(N+1)
        
def find(a):
    if a == parent[a]:
        return a
    else:
        parent[a] = find(parent[a])
        return parent[a]

def union(a, b):
    a = find(a)
    b = find(b)
    if a!=b:
        parent[b] = a

def checksame(a,b):
    a = find(a)
    b = find(b)
    if a==b:
        return True
    return False

for i in range(N+1):
    parent[i] = i

for _ in range(M):
    z, a, b = map(int, input().split())
    if z == 0:
        union(a,b)
        
    if z == 1:
        if checksame(a,b):
            print("YES")
        else:
            print("NO")

File: Java/백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());
        
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        
        int[] parent = new int[n+1];
        
        for(int i = 0; i < n+1; i++){
            parent[i] = i;
        }
        
        for(int i =0; i<m; i++){
            st = new StringTokenizer(br.readLine());
            int c = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            
            if(c == 0){
                union(parent, a, b);
            }
            
            else{
                if(checkSame(parent, a, b)){
                    sb.append("YES\n");
                }
                else{
                    sb.append("NO\n");
                }
            }    
        }
        
        System.out.print(sb);
    }
    
    static int find(int a, int[] parent){
        if(a == parent[a]){
            return a;
        }
        else{
            parent[a] = find(parent[a], parent);
            return parent[a];
        }
        
    }
    
    static void union(int[] parent, int a, int b){
        a = find(a, parent);
        b = find(b, parent);
        
        if(a != b){
            parent[b] = a;
        }
        
    }
    
    static boolean checkSame(int[] parent, int a, int b){
        a = find(a, parent);
        b = find(b, parent);
        
        if(a != b){
            return false;
        }
        else return true;
        
    }
}
